Bearded.Utilities/Core/Maybe.cs

```diff
@@ -21,6 +22,14 @@ private Maybe(T value)
 
         public T ValueOrDefault(T @default) => hasValue ? value : @default;
 
+        public T ValueOrDefault(Func<T> defaultProvider) => hasValue ? value : defaultProvider();
+
+        public Result<T, TError> ValueOrFailure<TError>(TError error) =>
+            hasValue ? (Result<T, TError>) Result.Success(value) : Result.Failure(error);
+
+        public Result<T, TError> ValueOrFailure<TError>(Func<TError> errorProvider) =>
```

Comment:
The name is not ideal since it does something quite different than ValueOrDefault, but I can't think of a better one either.

Reply:
What about `AsResult` to mimic the `AsMaybe` on `Result`?

Reply:
I see what you mean, but that makes the parameter not read very well/clearly perhaps?

```
someMaybe.AsResult(() => new MyException())
```

 feels pretty weird.

Reply:
I think my preference here is either `ValueOrFailure` or `AsResult`. I don't think I have a strong preference for either.

Reply:
Let's leave it as is.